### PR TITLE
feat: Support lazy initial state and double resolution bug fix

### PR DESCRIPTION
**Pull request by juan-foobar**
## Summary

This PR allows us to initialise state with a value set lazily. We do this by providing a callback rather than a value, the callback would then be called and the result of it would be the value used for the initial state.

We also fix a bug where the callback is attempted to be resolved twice.

Finally, the PR also refactors the `updater` to broaden its usage.

**Note**: _Aside from the bug fix, this PR is mostly a type change. You could previously "lazy initialise" a state since the `setGlobalState` function should only run once when the file loads. With the new changes, if you send a callback, it would be resolved in initialisation rather than you having to resolve it before setting it._

## How to test

Tests where included to show the support of this feature.

### CODE REVIEW

File: examples/05_typescript/src/Profile.tsx [hunk 1]
```diff
@@ -6,11 +6,12 @@ type Person = {
   age: number
 }
 
-const { useGlobalState } = createGlobalState<Person>({
+// We can also do lazy initialization
```
Comment (seblondono) [inline examples/05_typescript/src/Profile.tsx:9]: NIT: Is this comment really needed here?

File: src/createReadOnlyHook.ts [hunk 2]
```diff
@@ -6,22 +6,36 @@ export function createReadOnlyHook<TState, TPartial>(
   state$: Observable<TState>,
   project: (state: TState) => TPartial
 ) {
+
+  // We want to set the value as-is and avoid React resolving in case it is a function
+  /**
+   * If the value passed by the user is a function, we are already resolving it previously.
+   * Therefore, we want to set the value as is, and thus, we wrap it in a callback so React resolved to our value.
+   */
+  const generateProjectedValueCallback = (value: TState) => {
```
Comment (seblondono) [inline src/createReadOnlyHook.ts:15]: Do you think wrapping this on a `useCallback` would help perf in any way? I think it would not give a huge boost, but would not hurt either.

File: src/createReadOnlyHook.ts [hunk 3]
```diff
@@ -6,22 +6,36 @@ export function createReadOnlyHook<TState, TPartial>(
...
+       * We use the `initializationValue` and only update the state after mounting the component.
        * This helps us keep a consistent and predictable state between server and client rendering
        *  consequently avoiding a possible hydration mismatch.
        */
-       const stateValueAfterMount = project(state$.value)
-
-       if (state !== stateValueAfterMount) {
-         setState(stateValueAfterMount)
-       }
+      const stateValueAfterMountCallback = generateProjectedValueCallback(state$.value)
```
Comment (seblondono) [inline src/createReadOnlyHook.ts:32]: Are these 2 lines always to written one after the other? I mean, first `generateProjectedValueCallback` and the `setState`. If so, maybe wrap them in another function.

File: src/stateInputValueResolver.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+const isFunction = (value: unknown): value is Function => typeof value === 'function'
+
+const stateInputValueResolver = <CurrentValue, NextValue = CurrentValue>(
```
Comment (seblondono) [inline src/stateInputValueResolver.ts:3]: I am not sure if it is a standard to always prefix generic types with a `T`. I still think it would read better and make it double obvious we're talking about types here.

I would like an extra generic value for the type of the states value. I mean, for the `value` itself. This is because, the `currentValue`, the `prevState`, and `nextValue` should all have the same type. And I would also align the naming convention here, are we talking about `value`s or `state`s?

**Review by seblondono — CHANGES_REQUESTED** [review]

File: src/stateInputValueResolver.ts (hunk 4 above)
Comment (l-Leniac-l) [reply src/stateInputValueResolver.ts:3]: I agree with you. I would avoid using just `T` since the library already has a standard of using `TState` or `TValue`. But yes, let's make things more transparent on this function.

File: src/createReadOnlyHook.ts (hunk 3 above)
Comment (l-Leniac-l) [reply src/createReadOnlyHook.ts:32]: I don't think they are. But good input tho.

File: src/createReadOnlyHook.ts (hunk 2 above)
Comment (l-Leniac-l) [reply src/createReadOnlyHook.ts:15]: I don't think it is possible to use a callback here since this is not yet inside the hook. Should we move it inside the hook then? What's the trade-off of doing it?

File: examples/05_typescript/src/Profile.tsx (hunk 1 above)
Comment (l-Leniac-l) [reply examples/05_typescript/src/Profile.tsx:9]: Given this is an example of how to use it, I think it does not harm to have this comment.

Comment (seblondono) [reply examples/05_typescript/src/Profile.tsx:9]: Do you think it would be more complete if this was a JSDoc comment on `createGlobalState`? You don't have to do it now, but since this is also open source, I think that would be a nice addition to the library.

File: src/createReadOnlyHook.ts (hunk 2 above)
Comment (seblondono) [reply src/createReadOnlyHook.ts:15]: Ah I see, thanks for pointing that out. Does `generateProjectedValueCallback` need to live inside of `createReadOnlyHook` then? I just got a bit confused by it, because at the end what we are returning is just a function.

I got another question, tho, what does `project` mean in this case?

Comment (seblondono) [reply src/createReadOnlyHook.ts:15]: Let's have a call about this actually, I've got an idea but it's hard to describe here 😅

File: src/createReadOnlyHook.ts (hunk 3 above)
Comment (seblondono) [reply src/createReadOnlyHook.ts:32]: Well, I don't see any case where you would call `generateProjectedValueCallback` and not use the return to set the `state`... I think they do need to be called one after the other...

File: examples/05_typescript/src/Profile.tsx (hunk 1 above)
Comment (juan-foobar) [reply examples/05_typescript/src/Profile.tsx]: Agreed, I have removed the comment and added an [issue to reference this need](https://github.com/foobaragency/foo-state/issues/123).

File: src/createReadOnlyHook.ts (hunk 2 above)
Comment (juan-foobar) [reply src/createReadOnlyHook.ts:15]: After discussing this on a call, the following is what was concluded:
- The `generateProjectedValueCallback` function lives inside the `createReadOnlyHook` because:
  - `generateProjectedValueCallback` is only used by the `createReadOnlyHook` function,
  - the `createReadOnlyHook` should only be called on file loading which limits the `generateProjectedValueCallback` function to be defined once per "global state", and
  - we use the `project` function inside it without having to pass it through the props which in this case is acceptable since it is a parameter and therefore not expected to change within the `createReadOnlyHook` function.
- `generateProjectedValueCallback` is defined once on file load so memoizing the `generateProjectedValueCallback` function should not increase performance.
- `project` stands for _projection_ since it represents a "projection" of the value.

File: src/createReadOnlyHook.ts (hunk 3 above)
Comment (juan-foobar) [reply src/createReadOnlyHook.ts:32]: Although the state is set using the return value of the `generateProjectedValueCallback` function, the state is set differently. We use the return value of the `generateProjectedValueCallback` function to _initialise_ the state and later use the return value of the `generateProjectedValueCallback` function to _update_ the state.

This should not be abstracted in a function since they are doing different things. If we abstracted setting the state into a function, since it is being used twice, we might be over-abstracting and also opening the opportunity for diverging code which should do the same (generate a callback which resolves in the projected value).

File: src/stateInputValueResolver.ts (hunk 4 above)
Comment (juan-foobar) [reply src/stateInputValueResolver.ts]: >I am not sure if it is a standard to always prefix generic types with a T. I still think it would read better and make it double obvious we're talking about types here.

Thanks for the tip @seblondono, this makes much sense! 👌 I have renamed the types :v:

> I would like an extra generic value for the type of the states value. I mean, for the value itself. This is because, the currentValue, the prevState, and nextValue should all have the same type. And I would also align the naming convention here, are we talking about values or states?

We are talking about the values which represent the different stages of the state. We can have the value of the current state and the value of the next state, whereas the state is specifically what is saved.

When initialising the state the "current state" doesn't yet exist and therefore is `undefined`. In this case, the type of the "current state" and the type of the "next state" are different when initialising the state.

File: src/createReadOnlyHook.ts (hunk 3 above)
Comment (seblondono) [reply src/createReadOnlyHook.ts:32]: NIT: In line 10 and 11 there are 2 different types of comments. I think they kinda say the same thing, no? Would be cool if you can keep just one and maybe rephrase a bit so it is more clear, maybe something like
```
/**
* explanation
* if A case => B case
* else C case
* /
```

**Review by seblondono — APPROVED** [review]

File: src/createReadOnlyHook.ts (hunk 3 above)
Comment (juan-foobar) [reply src/createReadOnlyHook.ts:31]: Thanks! I completely missed this, I think it was a merging thing that went wrong or something 😅 

Removed the duplicated unnecessary comment ❤️
